### PR TITLE
Gateway Infrastructure API implementation

### PR DESCRIPTION
**Pull request by howardjohn**
**What type of PR is this?**

/kind gep

**What this PR does / why we need it**:
This PR implements GEP 1762, bumping it to experimental. As it depends on 1867, it also bumps that to experimental.

This introduces a new field, GatewayInfrastructure, to the experimental channel
**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
A new `infrastructure` field has been added to `Gateway`. Currently, this allows configuring labels and annotations to apply to underlying resources.
```

### CODE REVIEW

File: apis/v1beta1/gateway_types.go [hunk 1]
```diff
@@ -571,6 +579,30 @@ type GatewayStatus struct {
...
+	// Labels that should be applied to any resources created in response to this Gateway.
+	//
+	// For implementations creating other Kubernetes objects, this should be the `metadata.labels` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "labels" concepts.
+	//
+	// An implementation may chose to add additional implementation-specific labels as they see fit.
+	//
+	// Support: Extended
+	// +kubebuilder:validation:MaxItems=8
+	Labels map[string]string `json:"labels,omitempty"`
```
Comment (robscott) [inline apis/v1beta1/gateway_types.go:593]: Recommend copying the types we already use for TLS Option validation here:

https://github.com/kubernetes-sigs/gateway-api/blob/789326ba892891dcc92d05cae6b9487e63817aca/apis/v1beta1/gateway_types.go#L389

File: apis/v1beta1/gateway_types.go [hunk 2]
```diff
@@ -571,6 +579,30 @@ type GatewayStatus struct {
...
+type GatewayInfrastructure struct {
+	// Labels that should be applied to any resources created in response to this Gateway.
+	//
+	// For implementations creating other Kubernetes objects, this should be the `metadata.labels` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "labels" concepts.
+	//
+	// An implementation may chose to add additional implementation-specific labels as they see fit.
+	//
+	// Support: Extended
+	// +kubebuilder:validation:MaxItems=8
```
Comment (robscott) [inline apis/v1beta1/gateway_types.go:592]: Despite not giving any warning, this line is actually ineffective, we want MaxProperties. Example from TLS Options below (still agree with max of 8 in this case though):

https://github.com/kubernetes-sigs/gateway-api/blob/789326ba892891dcc92d05cae6b9487e63817aca/apis/v1beta1/gateway_types.go#L388

File: apis/v1beta1/gateway_types.go [hunk 3]
```diff
@@ -571,6 +579,30 @@ type GatewayStatus struct {
...
+	// Annotations that should be applied to any resources created in response to this Gateway.
+	//
+	// For implementations creating other Kubernetes objects, this should be the `metadata.annotations` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "annotations" concepts.
+	//
+	// An implementation may chose to add additional implementation-specific annotations as they see fit.
+	//
+	// Support: Extended
+	// +kubebuilder:validation:MaxItems=8
+	Annotations map[string]string `json:"annotations,omitempty"`
```
Comment (robscott) [inline apis/v1beta1/gateway_types.go:603]: Same comments as above

File: apis/v1beta1/gateway_types.go [hunk 4]
```diff
@@ -571,6 +579,30 @@ type GatewayStatus struct {
 	Listeners []ListenerStatus `json:"listeners,omitempty"`
 }
 
+// GatewayInfrastructure defines infrastructure level attributes about a Gateway instance.
+type GatewayInfrastructure struct {
+	// Labels that should be applied to any resources created in response to this Gateway.
```
Comment (robscott) [inline apis/v1beta1/gateway_types.go:584]:
```suggestion
	// Labels that SHOULD be applied to any resources created in response to this Gateway.
```

File: apis/v1beta1/gateway_types.go [hunk 5]
```diff
@@ -571,6 +579,30 @@ type GatewayStatus struct {
...
+	//
+	// For implementations creating other Kubernetes objects, this should be the `metadata.labels` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "labels" concepts.
+	//
+	// An implementation may chose to add additional implementation-specific labels as they see fit.
+	//
+	// Support: Extended
+	// +kubebuilder:validation:MaxItems=8
+	Labels map[string]string `json:"labels,omitempty"`
+	// Annotations that should be applied to any resources created in response to this Gateway.
```
Comment (robscott) [inline apis/v1beta1/gateway_types.go:594]:
```suggestion
	// Annotations that SHOULD be applied to any resources created in response to this Gateway.
```

Comment (robscott) [inline apis/v1beta1/gateway_types.go:594]: Very nitpicky, but prefer to have empty new line between fields..
```suggestion

	// Annotations that should be applied to any resources created in response to this Gateway.
```

**Review by robscott — COMMENTED** [review]
Thanks @howardjohn! This mostly LGTM, just a few nits.

**Comment by youngnick** [comment]
LGTM once @robscott's changes are done.

**Comment by robscott** [comment]
Thanks @howardjohn!

/lgtm
/approve

**Pull request merged**